### ISSUE TITLE
Wrong return type inference in conditional cases

### ISSUE THREAD

**Issue by awerlogus** [microsoft/TypeScript#38037]
**TypeScript Version:**  3.9.0-dev.20200418

**Search Terms:**
Return type
Return type inference
Return conditional

**Code**

When there's condition in return type, its generalization is based on 'else' type only 

```typescript
type F1 = <T>(arg: T) => T extends string ? T : 0

// Expected: string | 0
// Got: 0
type R1 = ReturnType<F1>
```
```typescript
type F2 = <T extends string>(arg: T) => T extends 'foo' ? T : 0

// Expected: 'foo' | 0
// Got: 0
type R2 = ReturnType<F2>
```

It gives us ability to produce bugs like

```typescript
declare function func<T extends string>(arg: T): T extends '0' ? 0 : T

// Works, but expected an error
const funcAlias: (arg: string) => string = func
```

**Comment by awerlogus** [comment]
One more example for parameters case:

```typescript
declare function func<T extends string>(arg: T, b: T extends '0' ? string : 0): T

// Got: [string, 0]
// Expected: [string, 0] | ['0', string]
type P = Parameters<typeof func>
```

**Comment by awerlogus** [comment]
Yet another example with sophisticated parameter type computation

```typescript
type AnyTuple = ReadonlyArray<unknown> & { readonly 0: unknown } | [] | readonly []

declare function func<T extends AnyTuple>(arg: T, b: T['length']): T

// Got: [AnyTuple, number]
// Expected: ???
type P = Parameters<typeof func>
```

If we had higher kinded types, it could look like

```typescript
type Params<T extends AnyTuple> = [T, T['length']]
```

But I'm not sure it's a good idea — to add HKT in TypeScript. It will produce even more extreme cases count to handle than we had before.

**Comment by awerlogus** [comment]
And my favorite: [Wrapping type checkers](https://dev.to/awerlogus/why-you-check-the-types-wrong-wrapping-type-checkers-ke)

```typescript
enum string_literal { string_literal = '' }

type StringLiteral<T> = T extends string ? string extends T ? string_literal : T : string_literal

declare function func<T>(arg: StringLiteral<T>): T

// Got: [string_literal] 
// Expected: ???
type P = Parameters<typeof func>
```

Looks right, but nothing is assignable to type string_literal because it works just as 'never' type

**Comment by ahejlsberg** [comment]
This is working as intended. In the example

```ts
type F1 = <T>(arg: T) => T extends string ? T : 0;

type R1 = ReturnType<F1>;  // 0
```

the `ReturnType<T>` type doesn't reason about higher-order types, so `F1` is instantiated with the constraint of `T` (i.e. `unknown`) before the condition is evaluated. It's similar to invoking `F1` with an argument of type `unknown`. If you want a different outcome for an argument that _might_ be of type `string`, you can add an additional case to the conditional type:

```ts
type F1 = <T>(arg: T) =>
    T extends string ? T :
    string extends T ? string | 0 :
    0;
```

The same reasoning holds for the other examples.

**Comment by awerlogus** [comment]
Got it, thanks.

**Comment by awerlogus** [comment]
I experimented a little and found this interesting example:

```ts
type F1 = <T>(arg: T) => T extends string ? string : 0

// Got: any
type R1 = ReturnType<F1>
```
It doesn't seem logical. Here `T` will be inferred as `unknown`, `unknown extends string` is `false`. So return type should be inferred as `0`. Here we got `any`.

**Comment by ahejlsberg** [comment]
It's a bit strange, but still working as intended. It helps to look at how `ReturnType<T>` is declared:

```ts
type ReturnType<T extends (...args: any) => any> = T extends (...args: any) => infer R ? R : any;
```

To resolve `ReturnType<F1>` we first infer from `F1` to `(...args: any) => R`. This causes us to infer `0` for `R`. We then check if `F1` is assignable to `(...args: any) => 0`. In order to do that, we instantiate `F1` in the context of `(...args: any) => 0` which produces `(arg: any) => string | 0` (because we inferred `any` for `T`). Since `(arg: any) => string | 0` is not assignable to `(...args: any) => 0`, the `ReturnType` conditional resolves to the false branch, which is `any`.

It all goes to prove that `ReturnType<T>` is rather useless with generic function types, which is a known restriction.

**Comment by awerlogus** [comment]
In my opinion, language design approach that generates a lot of such edge cases makes it completely impossible to create type safe library functions. If you add some feature you must be sure that it works fine with all existing language features, but not with just some of them. We need to do something with the current situation because complexity is keeping to grow. Already we have to handle:

1) Arrays of each type differently to return exact type of array that was passed:
```ts
type A<T extends ReadonlyArray<any>> = T extends Array<any> ? Array<...> : ReadonlyArray<...>
```

2) Never forget to add `| readonly []` to `ReadonlyArray<any>` if we want `T` might be inferred as a  tuple
```ts
type B<T extends ReadonlyArray<any> | readonly []> = ...
```

3) Then also handle tuples of each type differently to return as exact type as possible

```ts
type Tuple<T> = ReadonlyArray<T> & { readonly 0: T } | readonly [] // Maybe add this in ts lib?

type C<T extends ReadonlyArray<any> | readonly []> = T extends Array<any> 
  ? T extends Tuple<any> ? 'mutable tuple case' : 'mutable array case'
  : T extends Tuple<any> ? 'immutable tuple case' : 'immutable array case'
```

Fun fact: almost all people use `Array` as type constraint when they write functions that can work with tuples only:

```ts
type D<T extends Array<any>> = ...
```

4) But that's not all. T may be inferred as union of arrays or tuples and we need to handle it correctly too. So we can convert that union to tuple of its elements then handle each element and convert it all back to a union.

```ts
type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never

type UnionToFunctions<U> = U extends unknown ? (k: U) => void : never

type FunctionIntersectionToTuple<FNS> = _FunctionIntersectionToTuple<FNS, readonly []>

type _FunctionIntersectionToTuple<FNS, R extends Tuple<any>> =
  FNS extends ((arg: infer P) => void) ?
  FNS extends (((arg: P) => void) & infer P1) ? {
    0: _FunctionIntersectionToTuple<P1, [P, ...R]>
    1: [P, ...R]
  }[P1 extends (arg: any) => void ? 0 : 1] : never : never


/** Converts union to tuple */
export type UnionToTuple<U> = FunctionIntersectionToTuple<UnionToIntersection<UnionToFunctions<U>>>
```

```ts
type E<T extends ReadonlyArray<any> | readonly []> = HandleTupleByE<UnionToTuple<T>>

type HandleTupleByE<T extends Tuple<any>> = {
  0: T extends readonly [infer HEAD, ...infer TAIL]
  ? HEAD extends ReadonlyArray<any> | readonly []
  ? TAIL['length'] extends 0 ? C<HEAD>
  : TAIL extends Tuple<ReadonlyArray<any> | readonly []>
  ? C<HEAD> | HandleTupleByE<TAIL> : never : never : never
  1: never
}[T['length'] extends 0 ? 1 : 0]
```

```ts
// "mutable tuple case" | "immutable tuple case" | "immutable array case"
type F = E<readonly [1, 2] | [3, 4] | ReadonlyArray<number>>
```

Union elements are handled by type `C` from p.3. 

Let me remind you: this whole pile of code is used by a **single function**.

5) But also there're higher order functions that can accept our function as parameter. What if this function was written by a beginner who didn't know that `ReturnType` is useless for generics (there's no way to separate generic functions from non generic ones, so it is useless for all functions) and he used this? All our type checks will be converted to `any`.
But what if the function that we passed as a parameter was also overloaded? This case should be handled too. There're also higher order overloaded functions. It makes us write more and more type level code that will be completely broken by the first non standard use case because some language features are not compatible with others. I think this is an indicator of a sick language design.

I would be glad if you could put more effort into rethinking how to generalize and simplify the language rather than making it even more complex.
